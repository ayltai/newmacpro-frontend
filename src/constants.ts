import { Package, } from './types';

export const API_MAX_RETRIES           = 5;
export const HOMEBREW_REFRESH_INTERVAL = 1000 * 60 * 60 * 18;
export const TWEAKS_REFRESH_INTERVAL   = 1000 * 60 * 60 * 48;
export const BUNDLES_REFRESH_INTERVAL  = 1000 * 60 * 60 * 4;
export const MIN_SEARCH_LENGTH         = 2;
export const SEARCH_FUZZINESS          = 0.2;
export const SEARCH_MAX_RESULTS        = 50;

export const POPULAR_APPS : Package[] = [
    {
        id          : 'google-chrome',
        displayName : 'Google Chrome',
        description : 'Download the new Google Chrome for your iPhone and iPad. Now more simple, secure and faster than ever. Get the best of Google Search, and easily sync your bookmarks and passwords with Chrome on your laptop. Download the fast, secure browser recommended by Google.',
        source      : 'Homebrew/cask',
    },
    {
        id          : 'spotify',
        displayName : 'Spotify',
        description : 'With the Spotify music and podcast app, you can play millions of songs, albums and original podcasts for free. Stream music and podcasts, discover albums, playlists or even single songs for free on your mobile or tablet. Subscribe to Spotify Premium to download and listen offline wherever you are.',
        source      : 'Homebrew/cask',
    },
    {
        id          : 'alfred',
        displayName : 'Alfred',
        description : 'Alfred is an award-winning app for macOS which boosts your efficiency with hotkeys, keywords, text expansion and more. Search your Mac and the web, and be more productive with custom actions to control your Mac.',
        source      : 'Homebrew/cask',
    },
    {
        id          : 'visual-studio-code',
        displayName : 'Visual Studio Code',
        description : 'Visual Studio Code is a code editor redefined and optimized for building and debugging modern web and cloud applications.',
        source      : 'Homebrew/cask',
    },
    {
        id          : 'microsoft-office',
        displayName : 'Microsoft 365 for Mac',
        description : 'Office for Mac with Microsoft 365, gives you power and flexibility to get things done from virtually anywhere.',
        source      : 'Homebrew/cask',
    },
    {
        id          : 'notion',
        displayName : 'Notion',
        description : 'Notion is the connected workspace where better, faster work happens. Now with AI',
        source      : 'Homebrew/cask',
    },
    {
        id          : 'fantastical',
        displayName : 'Fantastical',
        description : 'Fantastical, the calendar app you won\'t be able to live without. Quickly create new events and reminders with natural language input and more.',
        source      : 'Homebrew/cask',
    },
    {
        id          : 'gimp',
        displayName : 'GIMP',
        description : 'GIMP is a cross-platform image editor available for GNU/Linux, macOS, Windows and more operating systems. It is free software, you can change its source code and distribute your changes.',
        source      : 'Homebrew/cask',
    },
    {
        id          : 'bartender',
        displayName : 'Bartender',
        description : 'Bartender is an award-winning app for macOS that superpowers your menu bar, giving you total control over your menu bar items, what\'s displayed, and when, with menu bar items only showing when you need them. Bartender improves your workflow with quick reveal, search, custom hotkeys and triggers, and lots more.',
        source      : 'Homebrew/cask',
    },
    {
        id          : '1password',
        displayName : '1Password',
        description : 'A password manager, digital vault, form filler and secure digital wallet. 1Password remembers all your passwords for you to help keep account information safe.',
        source      : 'Homebrew/cask',
    },
];

export const POPULAR_PACKAGES : Package[] = [
    {
        id          : 'node',
        displayName : 'Node.js',
        description : 'Node.jsÂ® is an open-source, cross-platform JavaScript runtime environment.',
        source      : 'Homebrew/core',
    },
    {
        id          : 'git',
        displayName : 'Git',
        description : 'Git is a free and open source distributed version control system designed to handle everything from small to very large projects with speed and efficiency. ',
        source      : 'Homebrew/core',
    },
    {
        id          : 'wget',
        displayName : 'Wget',
        description : 'GNU Wget is a free software package for retrieving files using HTTP, HTTPS, FTP and FTPS, the most widely used Internet protocols. It is a non-interactive commandline tool, so it may easily be called from scripts, cron jobs, terminals without X-Windows support, etc.',
        source      : 'Homebrew/core',
    },
    {
        id          : 'python3',
        displayName : 'Python 3',
        description : 'Python is a programming language that lets you work quickly and integrate systems more effectively.',
        source      : 'Homebrew/core',
    },
    {
        id          : 'awscli',
        displayName : 'AWS CLI',
        description : 'The AWS Command Line Interface (CLI) is a unified tool to manage your AWS services.',
        source      : 'Homebrew/core',
    },
    {
        id          : 'cmake',
        displayName : 'CMake',
        description : 'CMake is an open-source, cross-platform family of tools designed to build, test and package software.',
        source      : 'Homebrew/core',
    },
    {
        id          : 'maven',
        displayName : 'Maven',
        description : 'Maven is a build automation tool used primarily for Java projects.',
        source      : 'Homebrew/core',
    },
    {
        id          : 'jq',
        displayName : 'jq',
        description : 'jq is a lightweight and flexible command-line JSON processor.',
        source      : 'Homebrew/core',
    },
    {
        id          : 'docker',
        displayName : 'Docker',
        description : 'Docker is a set of platform as a service products that use OS-level virtualization to deliver software in packages called containers.',
        source      : 'Homebrew/core',
    },
    {
        id          : 'htop',
        displayName : 'htop',
        description : 'htop is an interactive process viewer for Unix systems.',
        source      : 'Homebrew/core',
    },
];

export const POPULAR_TWEAKS : Package[] = [
    {
        id          : 'tweak_macos_keep_windows',
        displayName : 'Always keep windows',
        description : 'macOS can remember all the tabs and windows you had open for the next time you launch it again.',
        source      : 'Tweak',
    },
    {
        id          : 'tweak_macos_show_hidden_files',
        displayName : 'Show hidden files in Finder',
        description : 'By default, macOS hides hidden files in Finder. This tweak disables this behavior.',
        source      : 'Tweak',
    },
    {
        id          : 'tweak_macos_disable_itunes_media_keys',
        displayName : 'Disable iTunes media keys',
        description : 'If you have iTunes installed, when you press Play on the keyboard, or connect the headset and Bluetooth speaker, iTunes will automatically run. This tweak disables this behavior.',
        source      : 'Tweak',
    },
    {
        id          : 'tweak_macos_subpixel_font_rendering',
        displayName : 'Change subpixel font rendering on external monitors',
        description : 'If you are running macOS on a Mac without a retina display, or with an external monitor that does not have an ultra-high resolution screen, you may have noticed that some fonts and text can appear as blurry and difficult to read. This is because macOS is rendering the text using subpixel font rendering.',
        source      : 'Tweak',
    },
    {
        id          : 'tweak_macos_disable_photos_auto_open',
        displayName : 'Disable Photos from opening automatically',
        description : 'macOS automatically opens Photos every time you insert a memory card or connect a device. This tweak disables this behavior.',
        source      : 'Tweak',
    },
    {
        id          : 'tweak_macos_show_file_extensions',
        displayName : 'Show file extensions in Finder',
        description : 'By default, macOS hides file extensions in Finder. This tweak disables this behavior.',
        source      : 'Tweak',
    },
    {
        id          : 'tweak_macos_disable_smart_quotes',
        displayName : 'Disable smart quotes',
        description : 'macOS automatically converts straight quotes to smart quotes. But for developers, it\'s annoying to have to manually change them back. This tweak disables this behavior.',
        source      : 'Tweak',
    },
    {
        id          : 'tweak_macos_disable_smart_dashes',
        displayName : 'Disable smart dashes',
        description : 'macOS automatically converts straight dashes to smart dashes. But for developers, it\'s annoying to have to manually change them back. This tweak disables this behavior.',
        source      : 'Tweak',
    },
];

type Document = {
    title   : string,
    docPath : string,
};

export const DOCUMENTS : Document[] = [
    {
        title   : 'footer.disclaimer',
        docPath : '/docs/disclaimer.md',
    },
    {
        title   : 'footer.privacy',
        docPath : '/docs/privacy.md',
    },
    {
        title   : 'footer.terms',
        docPath : '/docs/terms.md',
    },
];
