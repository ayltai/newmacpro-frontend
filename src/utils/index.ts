import * as Sentry from '@sentry/react';

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const capitalizeFirstLetter = (input : string) => input.charAt(0).toUpperCase() + input.slice(1);
