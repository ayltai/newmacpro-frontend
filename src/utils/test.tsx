import { ThemeProvider, } from '@mui/material';
import { configureStore, Store, } from '@reduxjs/toolkit';
import { act, render, } from '@testing-library/react';
import mediaQuery from 'css-mediaquery';
import { ConfirmProvider, } from 'material-ui-confirm';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { bundleApi, cartApi, } from '../apis';
import { POPULAR_APPS, } from '../constants';
import { cartReducer, prefReducer, store as reduxStore, } from '../states';
import { AppTheme, } from '../styles';
import { Cart, } from '../types';

export const mockApis = async (cart? : Cart) => {
    fetchMock.doMock(request => {
        if (request.url.endsWith('/api/bundles') && request.method === 'POST') return Promise.resolve({
            status : 201,
            body   : 'Created',
        });

        if (request.url.endsWith('/api/bundles/1') && request.method === 'GET') return Promise.resolve({
            status : 200,
            body   : JSON.stringify({
                id          : '1',
                displayName : 'Dummy',
                packages    : [
                    POPULAR_APPS[0],
                ],
            }),
        });

        if (request.url.endsWith('/api/bundles/1') && request.method === 'DELETE') return Promise.resolve({
            status : 200,
            body   : 'OK',
        });

        if (request.url.endsWith('/api/carts') && request.method === 'POST') return Promise.resolve({
            status : 201,
            body   : 'Created',
        });

        return Promise.resolve({
            status : 404,
            body   : 'Not found',
        });
    });

    await act(async () => {
        await reduxStore.dispatch(bundleApi.endpoints?.getBundle.initiate({
            bundleId : '1',
            idToken  : 'idToken',
        }));
        if (cart) await reduxStore.dispatch(cartApi.endpoints.addCart.initiate(cart));
    });
};

export const createMatchMedia = (width : number) => (query : string) => ({
    matches             : mediaQuery.match(query, {
        width,
    }),
    media               : '',
    onchange            : null,
    addListener         : () => {},
    removeListener      : () => {},
    addEventListener    : () => {},
    removeEventListener : () => {},
    dispatchEvent       : () => false,
});

export const createStore = (preloadedState? : Record<string, any>) => configureStore({
    preloadedState,
    reducer    : {
        cart                      : cartReducer,
        pref                      : prefReducer,
        [ bundleApi.reducerPath ] : bundleApi.reducer,
        [ cartApi.reducerPath   ] : cartApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(bundleApi.middleware, cartApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <ThemeProvider theme={AppTheme}>
                <ConfirmProvider>
                    {children}
                </ConfirmProvider>
            </ThemeProvider>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
