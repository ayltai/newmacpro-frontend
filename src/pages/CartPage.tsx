import { NotificationImportant as NotificationIcon, } from '@mui/icons-material';
import { Box, Divider, Link, Typography, useTheme, } from '@mui/material';
import React, { useEffect, useState, } from 'react';
import { useTranslation, } from 'react-i18next';
import { v4 as uuid, } from 'uuid';

import { useAddCartMutation, } from '../apis';
import { CopyableTextField, Page, PageContent, PageTitle, } from '../components';
import { useAppSelector, } from '../hooks';
import { EmptySection, GridViewSection, } from '../sections';
import { Package, } from '../types';
import { handleError, } from '../utils';
import { PageProps, } from './Page.types';

export const CartPage = ({
    open,
    onClose,
} : PageProps) => {
    const { packages, } : { packages : Package[], } = useAppSelector(state => state.cart);

    const [ originalPackages, setOriginalPackages, ] = useState<Package[]>(packages);
    const [ cartId,           setCartId,           ] = useState<string | undefined>();

    const [ addCart, addCartResult, ] = useAddCartMutation();

    const theme = useTheme();

    const { t, } = useTranslation();

    useEffect(() => setCartId(uuid()),             [ open, ]);
    useEffect(() => setOriginalPackages(packages), [ open, ]);

    useEffect(() => {
        if (open && cartId && packages.length > 0) {
            addCart({
                id : cartId,
                packages,
            }).catch(handleError);
        }
    }, [ open, cartId, packages, ]);

    useEffect(() => {
        if (addCartResult.error) handleError(addCartResult.error);
    }, [ addCartResult.error, ]);

    return (
        <Page
            open={open}
            onClose={onClose}>
            <PageTitle>
                {t('cart.title', {
                    count : packages.length,
                })}
            </PageTitle>
            <PageContent>
                {packages.length > 0 && (
                    <>
                        <GridViewSection
                            title={t('cart.list_title')}
                            maxRows={2}
                            packages={originalPackages} />
                        <Divider />
                        {packages.some(pkg => pkg.source === 'App Store') && (
                            <Box
                                marginY={2}
                                display='flex'
                                alignItems='center'>
                                <NotificationIcon />
                                <Link
                                    marginLeft={1}
                                    href='https://support.apple.com/guide/app-store/fir6253293d/mac'
                                    target='_blank'>
                                    {t('note.app_store')}
                                </Link>
                            </Box>
                        )}
                        {packages.some(pkg => pkg.source === 'Tweak') && (
                            <Box
                                marginY={2}
                                display='flex'
                                alignItems='center'>
                                <NotificationIcon />
                                <Link
                                    marginLeft={1}
                                    href='https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection'
                                    target='_blank'>
                                    {t('note.tweak')}
                                </Link>
                            </Box>
                        )}
                        <Divider />
                        <Typography
                            marginTop={2}
                            gutterBottom>
                            {t('cart.instruction')}
                        </Typography>
                        <CopyableTextField
                            sx={{
                                backgroundColor : theme.palette.divider,
                            }}
                            fullWidth
                            copyHint={t('cart.copy.success')}
                            copyText={t('cart.copy.action')}
                            value={`/bin/bash -c "$(curl -fsSL '${process.env.NODE_ENV === 'development' ? 'http://localhost:3001' : process.env.REACT_APP_BACKEND_ENDPOINT!}/api/carts/${cartId}')"`} />
                    </>
                )}
                {packages.length === 0 && (
                    <EmptySection>{t('cart.empty')}</EmptySection>
                )}
            </PageContent>
        </Page>
    );
};
