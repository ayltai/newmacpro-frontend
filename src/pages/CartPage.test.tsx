import React from 'react';

import { POPULAR_APPS, POPULAR_TWEAKS, } from '../constants';
import { mockApis, render, waitFor, } from '../utils/test';
import { CartPage, } from './CartPage';

describe('<CartPage />', () => {
    it('renders correctly', async () => {
        await mockApis({
            id       : '1',
            packages : [
                POPULAR_APPS[0],
            ],
        });

        const component = render(<CartPage open />, {
            preloadedState : {
                cart : {
                    packages : [
                        POPULAR_APPS[0],
                    ],
                },
                pref : {
                    filters : [],
                },
            },
        });

        await waitFor(() => expect(component).toMatchSnapshot());
    });

    it('renders note for App Store packages', async () => {
        await mockApis({
            id       : '1',
            packages : [
                {
                    ...POPULAR_APPS[0],
                    source : 'App Store',
                },
            ],
        });

        const { getAllByRole, } = render(<CartPage open />, {
            preloadedState : {
                cart : {
                    packages : [
                        {
                            ...POPULAR_APPS[0],
                            source : 'App Store',
                        },
                    ],
                },
                pref : {
                    filters : [],
                },
            },
        });

        await waitFor(() => {
            expect(getAllByRole('link')[0]).toHaveTextContent('note.app_store');
        });
    });

    it('renders note for tweaks', async () => {
        await mockApis({
            id       : '1',
            packages : [
                POPULAR_TWEAKS[0],
            ],
        });

        const { getAllByRole, } = render(<CartPage open />, {
            preloadedState : {
                cart : {
                    packages : [
                        POPULAR_TWEAKS[0],
                    ],
                },
                pref : {
                    filters : [],
                },
            },
        });

        await waitFor(() => {
            expect(getAllByRole('link')[0]).toHaveTextContent('note.tweak');
        });
    });

    it('does not render any notes', async () => {
        await mockApis({
            id       : '1',
            packages : [
                POPULAR_APPS[0],
            ],
        });

        const { queryAllByRole, } = render(<CartPage open />, {
            preloadedState : {
                cart : {
                    packages : [
                        POPULAR_APPS[0],
                    ],
                },
                pref : {
                    filters : [],
                },
            },
        });

        await waitFor(() => {
            expect(queryAllByRole('link')).toHaveLength(0);
        });
    });
});
