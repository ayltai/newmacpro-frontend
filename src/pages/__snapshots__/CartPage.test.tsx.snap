// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartPage /> renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-sox5kk-MuiBackdrop-root"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
    </div>
    <div
      class="MuiModal-root MuiDialog-root css-29ydf2-MuiModal-root-MuiDialog-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-1sep8xo-MuiDialog-container"
        role="presentation"
        style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby=":r0:"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullScreen css-hlf5l-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-jp4g6a-MuiTypography-root-MuiDialogTitle-root"
            id=":r0:"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1ywgf30-MuiTypography-root"
            >
              cart.title
            </h6>
          </h2>
          <div
            class="MuiDialogContent-root MuiDialogContent-dividers css-1t4vnk2-MuiDialogContent-root"
          >
            <div
              class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
            >
              <div
                class="MuiBox-root css-0"
              >
                <div
                  class="MuiBox-root css-1n651cb"
                >
                  <div
                    class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
                  >
                    <div
                      class="MuiStack-root css-5andtu-MuiStack-root"
                    >
                      <h4
                        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-u2yth7-MuiTypography-root"
                      >
                        cart.list_title
                      </h4>
                      <div
                        class="MuiBox-root css-0"
                      >
                        <div
                          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1tkc62k-MuiPaper-root-MuiAccordion-root"
                        >
                          <div
                            aria-expanded="false"
                            class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
                            role="button"
                            tabindex="0"
                          >
                            <div
                              class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
                            >
                              <div
                                class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-spacing-xs-2 css-1psjskm-MuiGrid2-root"
                              >
                                <div
                                  class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-sm-6 MuiGrid2-grid-md-4 MuiGrid2-grid-lg-3 css-12xj993-MuiGrid2-root"
                                >
                                  <div
                                    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-f6z5rr-MuiPaper-root-MuiCard-root"
                                  >
                                    <div
                                      class="MuiBox-root css-1sm2s1z"
                                    >
                                      <div
                                        class="MuiBox-root css-1rr4qq7"
                                      >
                                        <button
                                          class="MuiButtonBase-root MuiCardActionArea-root css-10pclym-MuiButtonBase-root-MuiCardActionArea-root"
                                          tabindex="0"
                                          type="button"
                                        >
                                          <div
                                            class="MuiBox-root css-4rmlxi"
                                          >
                                            <div
                                              class="MuiCardContent-root css-1qk9xnj-MuiCardContent-root"
                                            >
                                              <h6
                                                class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1ywgf30-MuiTypography-root"
                                                title="Google Chrome"
                                              >
                                                Google Chrome
                                              </h6>
                                              <p
                                                class="MuiTypography-root MuiTypography-body1 css-1qnobji-MuiTypography-root"
                                                title="Download the new Google Chrome for your iPhone and iPad. Now more simple, secure and faster than ever. Get the best of Google Search, and easily sync your bookmarks and passwords with Chrome on your laptop. Download the fast, secure browser recommended by Google."
                                              >
                                                Download the new Google Chrome for your iPhone and iPad. Now more simple, secure and faster than ever. Get the best of Google Search, and easily sync your bookmarks and passwords with Chrome on your laptop. Download the fast, secure browser recommended by Google.
                                              </p>
                                              <div
                                                class="MuiBox-root css-13brihr"
                                              />
                                            </div>
                                          </div>
                                          <span
                                            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
                                          />
                                          <span
                                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                                          />
                                        </button>
                                      </div>
                                      <div
                                        class="MuiCardActions-root MuiCardActions-spacing css-11leuus-MuiCardActions-root"
                                      >
                                        <button
                                          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeSmall MuiButton-textSizeSmall css-1ujgr8z-MuiButtonBase-root-MuiButton-root"
                                          tabindex="0"
                                          type="button"
                                        >
                                          package.more_info
                                          <span
                                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                                          />
                                        </button>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
                            style="min-height: 0px;"
                          >
                            <div
                              class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                            >
                              <div
                                class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                              >
                                <div
                                  class="MuiAccordion-region"
                                  role="region"
                                />
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <hr
                class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
              />
              <hr
                class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
              />
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-1bsgx50-MuiTypography-root"
              >
                cart.instruction
              </p>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedEnd css-17je5yw-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                  type="text"
                  value="/bin/bash -c "$(curl -fsSL 'undefined/api/carts/1')""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
                >
                  <button
                    aria-label="cart.copy.action"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                    data-mui-internal-clone-element="true"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="ContentCopyIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-sox5kk-MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
